tests/components/test_topic_calibration.py

```diff
@@ -23,13 +23,11 @@ def test_request_with_topic_calibrator():
         ArticleSet(articles=req.todays_articles),
         ArticleSet(articles=req.past_articles),
         req.interest_profile,
-        req.num_recs,
```

Comment:
When we were building pipelines on the fly in the request/response cycle, we could configure the number of recs with a request parameter. Now that pipelines are built ahead and cached, the pipeline and component APIs as currently designed don't make that easy. We might want to make further changes to re-enable this functionality, but for now we can get away with making all recs pipelines return 10 results.

Reply:
IMO the easiest way to handle this will be to make the recommendation list length a runtime input instead of a configuration parameter, and wire it up as an input.

With one additional change to the new pipeline, this can even be semi-automated.

Reply:
There are reasons why this doesn't work super-well in some cases that I encountered in the Haystack world and were really a pain. When you define a pipeline with multiple candidate sources that need to fetch more candidates than the pipeline will eventually output, making the list length dynamic causes dilemmas and headaches about how to adjust the number of retrieved candidates relative to the list length.

To make that concrete, for an eventual list length of 10 items:
* Candidate source A returns 25 items
* Candidate source B returns 25 items
Then there's some ranker that takes the combined list of candidates and picks the best ones. Now you want to evaluate a metric at different cut-offs, so the number of results returned by the last component of the pipeline needs to change.

The first hurdle you have to get over is "how do I change the number of results from all the parts of the pipeline that need to be adjusted to make that work but not the ones that don't? how do I know which ones are which?" If you change the number of eventual results returned to 5, you can do that just by truncating the list (but that not be representative of what you think you're measuring if you're still picking from 50 candidates.) If you change it to 100, then there aren't enough candidates to do that so upstream components have to get adjusted too.

Haystack ended up in a situation where you could dynamically adjust how many results came back, but it would set the number of results _from every component_ to that number, which is not at all obvious or what you'd like to have happen in this kind of scenario. That's what happens when you try to make list length dynamic without thinking carefully about it though.

Reply:
(I'm not saying that it's an unsolvable problem but I've hesitated to make the list length dynamic on account of that unexpected complexity)

Reply:
That makes sense. There are some… interesting… things you can do here with the pipeline that can encode some relevant policies, I don't know if they're a *good* idea but they're fun to think about.

For example, if you want to run with "candidate lists are 5x the target rec list depth", we could do:

```python
n_recs = pipe.create_input('n_recs', int)
n_cands = pipe.add_component(multiply, a=n_recs, b=5)

cand = pipe.add_component(CandidateSelector, n=n_cands)
score = pipe.add_component(ItemScorer, items=cand)
recs = pipe.add_component(FancyRanker, items=score, n=n_recs)
```

Then `pipe.run(..., n_recs=5)` will use 25 candidates, and `pipe.run(..., n_recs=10)` will use 50.

But also hard-wiring the list length(s) makes sense; long-term it might be worth considering mechanisms to reuse unchanged components of an execution graph measuring a pipeline with different target list lengths.

Reply:
Yeah, we should think carefully about this—it seems like it ought to be solvable but half-baked solutions might be worse than nothing (as we discovered evaluating Haystack pipelines.)

Reply:
Yeah. I'm inclined to make sure that the core abstraction doesn't paint us into too much of one corner, and we can use practical experience to refine it.

I've got another idea for making the pipelines reconfigurable too, that will be another PR:

* add a `how="reuse"` option to `clone` that reuses existing component interfaces and only clones the pipeline structure and wiring (already planning to do this)
* instantiate and cache a "default" pipeline
* pipelines with non-default list lengths can be created by calling `clone(how="reuse")` on the cached pipeline, and then using `replace_component` to replace the components that depend on list length with reconfigured ones. This depends on those components being lightweight, but I think that's the common case.

If the last step here is common enough, we could also add a `reconfigure_component` method that works like `replace_component` (types and error handling omitted for brevity):

```python
def reconfigure_component(self, node, **changes):
    comp = node.component
    cfg = comp.get_config()
    cfg.update(changes)
    c2 = comp.from_config(cfg)
    self.replace_node(node, c2, <wiring>)
```

Reply:
This has gotten complicated enough that I'm starting to be worried that either something is off about the abstractions or we're building way more functionality than we actually need 😅

Reply:
Yeah. Let's use the simple "bake it in" design for now and figure out how to support reconfigurability for POPROX when we actually need it.